File: src/elicit/confirm.ts
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import type { SqlCategory } from '../types.js';

type ServerHandle = McpServer['server'];

export type GrantChoice = 'once' | 'session' | 'decline';

/**
 * Outcome of asking the user to authorize a statement.
 *
 * `unavailable` means the question was never put to them - the client does not
 * support elicitation, or the request failed. It is deliberately not folded
 * into `decline`, so callers can say what actually happened.
 */
export type ConfirmOutcome =
  | { choice: GrantChoice }
  | { choice: 'unavailable'; reason: string };

const CATEGORY_LABEL: Record<SqlCategory, string> = {
  read: 'read',
  write: 'WRITE',
  ddl: 'DDL (schema-changing)',
  admin: 'ADMIN',
  txCtrl: 'transaction control',
};

const GRANT_CHOICES: readonly GrantChoice[] = ['once', 'session', 'decline'];

function isGrantChoice(value: unknown): value is GrantChoice {
  return typeof value === 'string' && (GRANT_CHOICES as readonly string[]).includes(value);
}

function describeError(error: unknown): string {
  if (error instanceof Error) {
    return error.message;
  }
  return typeof error === 'string' ? error : 'unknown error';
}

export function makeConfirmFn(server: ServerHandle) {
  return async (args: {
    category: SqlCategory;
    statement: string;
    connectionName: string;
    database?: string | null;
  }): Promise<ConfirmOutcome> => {
    const snippet =
      args.statement.length > 800
        ? `${args.statement.slice(0, 800)}…`
        : args.statement;

    const label = CATEGORY_LABEL[args.category];

    const target = args.database
      ? `${args.connectionName} · ${args.database}`
      : args.connectionName;
    const sessionScope = args.database
      ? `all ${label} statements on ${args.database} until the MCP server restarts`
      : `all ${label} statements on this connection until the MCP server restarts`;

    // Elicitation is an optional MCP capability. Asking a client that never
    // declared it produces a "method not found" error that is indistinguishable
    // from a refusal, so check first and report the real reason.
    let capabilities;
    try {
      capabilities = server.getClientCapabilities();
    } catch (error) {
      return { choice: 'unavailable', reason: describeError(error) };
    }

    if (!capabilities?.elicitation) {
      return {
        choice: 'unavailable',
        reason: 'this MCP client does not support elicitation (no prompt can be shown)',
      };
    }

    try {
      const result = await server.elicitInput({
        message:
          `About to run a ${label} statement on ${target}.\n\n` +
          `--- SQL ---\n${snippet}\n--- end ---\n\n` +
          `Pick an authorization scope. "Allow for session" skips the prompt for ${sessionScope} ` +
          `To make this permanent, use the set_database_policy tool.`,
        requestedSchema: {
          type: 'object',
          properties: {
            choice: {
              type: 'string',
              title: 'Authorization',
              description: 'How should this statement be authorized?',
              enum: ['once', 'session', 'decline'],
              enumNames: [
                'Allow once (this statement only)',
                `Allow for session (${sessionScope})`,
                'Decline',
              ],
            },
          },
          required: ['choice'],
        },
      });

      // Per the MCP spec, "cancel" means the client dismissed the request
      // *without* an explicit choice - a timeout, no UI to show it in, the
      // user closing a dialog unanswered. That is not the same thing as
      // "decline" (an explicit no) and must not be reported as one: a client
      // that can never get a real answer (e.g. running in an unattended
      // mode with no interactive surface) will legitimately resolve every
      // elicitation this way, and folding it into "decline" fabricates a
      // refusal nobody made.
      if (result.action === 'cancel') {
        return {
          choice: 'unavailable',
          reason:
            'the client dismissed the prompt without an explicit choice - no UI to show it in, a timeout, or it was closed unanswered',
        };
      }

      // Only "decline" left as a non-accept action; it is a real answer.
      if (result.action !== 'accept') return { choice: 'decline' };

      const value = result.content?.['choice'];
      return { choice: isGrantChoice(value) ? value : 'decline' };
    } catch (error) {
      return { choice: 'unavailable', reason: describeError(error) };
    }
  };
}

File: src/mcp/confirm.rs
//! Elicitation bridge: connects the sync gate (run on the blocking pool)
//! to the async rmcp elicitation round trip.

use crate::approval::{ConfirmOutcome, GrantChoice};
use rmcp::schemars;
use std::sync::mpsc as std_mpsc;

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct ChoiceForm {
    /// How to authorize the statement: once, session, or decline.
    choice: GrantChoice,
}

rmcp::elicit_safe!(ChoiceForm);

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct PasswordForm {
    /// The MySQL/MariaDB password. Stored locally in the macOS Keychain;
    /// never in tool arguments, logs, or the audit DB.
    password: String,
}

rmcp::elicit_safe!(PasswordForm);

/// A sink whose `confirm` blocks on a channel that the async side answers
/// via elicitation. Built per request.
pub struct ElicitationSink {
    tx: std_mpsc::SyncSender<ElicitAsk>,
}

pub enum ElicitAsk {
    Request {
        message: String,
        reply: tokio::sync::oneshot::Sender<ConfirmOutcome>,
    },
}

impl ElicitationSink {
    pub fn new(tx: std_mpsc::SyncSender<ElicitAsk>) -> Self {
        Self { tx }
    }
}

impl crate::app::gate::ApprovalSink for ElicitationSink {
    fn confirm(&self, request: crate::app::gate::ApprovalRequest) -> ConfirmOutcome {
        use crate::policy::model::SqlCategory;
        let label = match request.category {
            SqlCategory::Read => "read",
            SqlCategory::Write => "WRITE",
            SqlCategory::Ddl => "DDL (schema-changing)",
            SqlCategory::Admin => "ADMIN",
            SqlCategory::TxCtrl => "transaction control",
        };
        let target = match &request.database {
            Some(db) => format!("{} · {}", request.connection_name, db),
            None => request.connection_name.clone(),
        };
        let tables: Vec<String> = request
            .tables
            .iter()
            .map(|t| format!("{}.{}", t.database, t.table))
            .collect();
        let session_scope = if request.database.is_some() {
            format!(
                "all {label} statements on {} until the MCP server restarts",
                request.database.clone().unwrap_or_default()
            )
        } else {
            format!("all {label} statements on this connection until the MCP server restarts")
        };
        let message = format!(
            "About to run a {label} statement on {target}.\n\n--- SQL ---\n{}\n--- end ---\n\nAffected tables: {}\n\nPick an authorization scope. \"Allow for session\" skips the prompt for {session_scope}. To make this permanent, use the set_table_policy tool.",
            request.statement_snippet,
            if tables.is_empty() {
                "(statement scope)".to_string()
            } else {
                tables.join(", ")
            }
        );
        let (reply_tx, reply_rx) = tokio::sync::oneshot::channel();
        if self
            .tx
            .send(ElicitAsk::Request {
                message,
                reply: reply_tx,
            })
            .is_err()
        {
            return ConfirmOutcome::Unavailable {
                reason: "approval channel closed".into(),
            };
        }
        // Block until the async side resolves (this runs on the blocking
        // pool, not the async runtime).
        match reply_rx.blocking_recv() {
            Ok(outcome) => outcome,
            Err(_) => ConfirmOutcome::Unavailable {
                reason: "approval channel dropped".into(),
            },
        }
    }
}

/// Drive one password-capture elicitation round trip for
/// `add_connection`. Accept-with-nonempty-string ⇒ `Ok(password)`
/// (zeroized on drop); decline/cancel/unavailable ⇒ `Err(reason)` — the
/// caller must NOT save the connection in that case.
pub async fn run_password_elicitation(
    peer: &rmcp::service::Peer<rmcp::RoleServer>,
    message: String,
) -> Result<zeroize::Zeroizing<String>, String> {
    match peer.elicit::<PasswordForm>(message).await {
        Ok(Some(form)) if !form.password.is_empty() => Ok(zeroize::Zeroizing::new(form.password)),
        Ok(Some(_)) => Err("the client accepted but returned an empty password".into()),
        Ok(None) => Err("the client accepted but returned no password content".into()),
        Err(rmcp::service::ElicitationError::UserDeclined) => {
            Err("password capture declined".into())
        }
        Err(rmcp::service::ElicitationError::UserCancelled) => {
            Err("password capture cancelled".into())
        }
        Err(e) => Err(format!("elicitation unavailable: {e}")),
    }
}

/// Drive one elicitation round trip on the async side. Mirrors the legacy
/// semantics exactly: cancel and empty/malformed content are `unavailable`,
/// never a decline.
pub async fn run_elicitation(
    peer: &rmcp::service::Peer<rmcp::RoleServer>,
    message: String,
) -> ConfirmOutcome {
    match peer.elicit::<ChoiceForm>(message).await {
        Ok(Some(form)) => ConfirmOutcome::Chosen(form.choice),
        Ok(None) => ConfirmOutcome::Unavailable {
            reason: "the client accepted but returned no choice content".into(),
        },
        Err(rmcp::service::ElicitationError::UserDeclined) => {
            ConfirmOutcome::Chosen(GrantChoice::Decline)
        }
        Err(rmcp::service::ElicitationError::UserCancelled) => ConfirmOutcome::Unavailable {
            reason: "the client dismissed the prompt without an explicit choice".into(),
        },
        Err(e) => ConfirmOutcome::Unavailable {
            reason: format!("elicitation failed: {e}"),
        },
    }
}
